lib/elixir/lib/calendar/duration.ex

```diff
@@ -355,6 +355,114 @@ defmodule Duration do
     end
   end
 
+  @doc """
+  Converts the given `duration` to a human readable representation.
+
+  ## Options
+
+    * `:units` - the units to be used alongside each duration component.
+      The default units follow the ISO 80000-3 standard:
+
+          [
+            year: "a",
+            month: "mo",
+            week: "wk",
+            day: "d",
+            hour: "h",
+            minute: "min",
+            second: "s"
+          ]
+
+    * `:separator` - a string used to separate the distinct components. Defaults to `" "`.
+
+  ## Examples
+
+      iex> Duration.to_string(Duration.new!(second: 30))
+      "30s"
+      iex> Duration.to_string(Duration.new!(day: 40, hour: 12, minute: 42, second: 12))
+      "40d 12h 42min 12s"
+
+  By default, this function uses ISO 80000-3 units, which uses "a" for years.
+  But you can customize all units via the units option:
+
+      iex> Duration.to_string(Duration.new!(year: 3))
+      "3a"
+      iex> Duration.to_string(Duration.new!(year: 3), units: [year: "y"])
+      "3y"
+
+  You may also choose the separator:
+
+      iex> Duration.to_string(Duration.new!(day: 40, hour: 12, minute: 42, second: 12), separator: ", ")
+      "40d, 12h, 42min, 12s"
+
+  A duration without components is rendered as "0s":
+
+      iex> Duration.to_string(Duration.new!([]))
+      "0s"
+
+  Microseconds are rendered as part of seconds with the appropriate precision:
+
+      iex> Duration.to_string(Duration.new!(second: 1, microsecond: {2_200, 3}))
+      "1.002s"
+      iex> Duration.to_string(Duration.new!(second: 1, microsecond: {-1_200_000, 4}))
+      "-0.2000s"
```

Comment:
Negative numbers are tricky with durations. Currently we have:

```elixir
iex> Duration.to_string(Duration.new!(hour: -1, minute: 30))
"-1h 30min"
iex> Duration.to_string(Duration.new!(hour: -1, minute: -30))
"-1h -30min"
```

Maybe I'm thinking about this all wrong but I'd expect:

```elixir
iex> Duration.to_string(Duration.new!(hour: -1, minute: 30))
"-30min"
iex> Duration.to_string(Duration.new!(hour: -1, minute: -30))
"-1h 30min"
```

Reply:
You are right.

As of  now the Duration type is:

```Elixir
@type t() :: %Duration{
  day: integer(),
  hour: integer(),
  microsecond: {integer(), 0..6},
  minute: integer(),
  month: integer(),
  second: integer(),
  week: integer(),
  year: integer()
}
```

maybe the struct should be something like:

```elixir
@type t() :: %Duration{
  sign: 1 | -1,
  day: non_neg_integer(),
  hour: non_neg_integer(),
  microsecond: {non_neg_integer(), 0..6},
  minute: non_neg_integer(),
  month: non_neg_integer(),
  second: non_neg_integer(),
  week: non_neg_integer(),
  year: non_neg_integer()
}
```

With a new field called `sign`.

Reply:
imho this is ok. I wasn't able to find this really specified in ISO 8000 but it aligns with ISO 8601 durations. I get that it stands out in terms of "humanly readable" _but_ i'd prefer the string to represent all units in the Duration as they are, over having the string representation calculate them away. How about prefixing everything, so that:

```
iex> Duration.to_string(Duration.new!(hour: -1, minute: 30))
"-1h +30min"
```

Reply:
This is correct as is. You can't express `12 months  - 1 week` if we only had a sign field.

Reply:
And similarly, we can't resolve `12 months - 1 week` by breaking months into a number of weeks.

Reply:
Thank you for the clarification. I was not in touch when this feature when was introduced. I see that some parts of it can be positive, some negative. That's interesting.